---
title: "Post-processing of scRNA-seq feature barcode matrix data using Seurat"
author: "Courtney Wong"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    df_print: paged
    number_sections: true
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(
	echo = TRUE
)
```

```{r lib}
suppressPackageStartupMessages({
  library(dplyr)
  library(Seurat)
  library(patchwork)
  library(ggplot2)
})
```

# Background

The sample selected for interrogation below is BALF-C141 from Liao et al. (2020), bronchoalveolar immune cells taken from the lung of a patient with mild COVID-19. Feature Barcode Matrices in HDF5 format were generated from FASTQ files using CellRanger and imported below. Note that we use the filtered matrix, where only barcodes believed to be associated with cells are retained, and "background barcodes" are removed.

As this is part of a reproducibility study, our aim is to follow the steps taken by Liao et al. (2020), as described in the sample metadata on [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4339769), while taking note of how these steps may have impacted the final result.

# Load data

Below, we load the filtered .h5 gene count data generated by CellRanger and store it into a Seurat object. Note that we could have set a threshold for `min.cells` (features were detected in at least this many cells) and `min.features` (cells where at least this many features were detected), but Liao et al. (2020) did not mention deviating from the default for these parameters.

```{r}
# Load the filtered dataset generated by CellRanger
BALF_C141.data <- Read10X_h5("filtered_feature_bc_matrix.h5")
# Initialize Seurat object with raw data
BALF_C141 <- CreateSeuratObject(counts = BALF_C141.data,
                                project = "covid")
```

# QC and Cell Filtering

Next, we assess the quality of the cells based on the number of unique features/genes detected (`nFeature_RNA`, extremely low gene count suggests empty droplets or poor quality cells while extremely high gene count suggests doublets) and the percentage of reads that map to mitochondrial genes (`percent.mt`, a high percentage suggests contamination due to low-quality or dying cells). `nCount_RNA` is the total number of molecules detected in the cell.

```{r}
# Calculate percentage of genes that are mitochondrial
BALF_C141[["percent.mt"]] <- PercentageFeatureSet(BALF_C141, pattern = "^MT-")
VlnPlot(BALF_C141, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0.0001)
# Create scatterplots to view trend
FeatureScatter(BALF_C141, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + 
  theme(legend.position="none")
FeatureScatter(BALF_C141, feature1 = "nCount_RNA", feature2 = "percent.mt") + 
  theme(legend.position="none")
```

Liao et al. (2020) filtered the cells using the following criteria: "gene number between 200 and 6000, UMI count above 1000 and mitochondrial gene percentage below 0.1." We can reason that this will eliminate the following:

-   the cluster of cells in the first violin plot with extremely low feature count
-   the cluster of cells in the third violin plot with extremely high (~90%) mitochondrial mRNA
-   roughly, the upper tail observed in the nCount vs nFeature graph

We apply this feature below using `subset()` then re-plot the same figures.

```{r}
BALF_C141 <- subset(BALF_C141, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA > 1000 & percent.mt < 10)
VlnPlot(BALF_C141, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0.0001)
FeatureScatter(BALF_C141, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + 
  theme(legend.position="none")
FeatureScatter(BALF_C141, feature1 = "nCount_RNA", feature2 = "percent.mt") + 
  theme(legend.position="none")
```

# Normalization

Liao et al. (2020) normalizes the filtered gene-barcode matrix using the LogNormalize method.

```{r}
BALF_C141 <- NormalizeData(BALF_C141, normalization.method = "LogNormalize")
```

# Selection of Highly Variable Genes

Liao et al. (2020) then "analyzed by principal component analysis (PCA) using the top 2,000 most variable genes." They do not mention a selection method for choosing the top variable features, so we use the default, `vst`, which standardizes feature values using the observed mean and expected variance predicted by local polynomial regression (loess) between log(variance) and log(mean).

```{r}
BALF_C141 <- FindVariableFeatures(BALF_C141, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
p <- VariableFeaturePlot(BALF_C141) + theme(legend.position = "bottom")
LabelPoints(plot = p,
            points = head(VariableFeatures(BALF_C141), 10),
            repel = TRUE)
```

# Scaling



# Dimensionality Reduction with PCA



# References

Liao, M., Liu, Y., Yuan, J. et al. Single-cell landscape of bronchoalveolar immune cells in patients with COVID-19. Nat Med 26, 842â€“844 (2020). https://doi.org/10.1038/s41591-020-0901-9